### ISSUE TITLE
Runtime issues caused by assets directory

### ISSUE THREAD

**Issue by orhun** [fishfight/FishFight#296]
An user (@woelfman) on Reddit pointed out some issues they encountered about running FishFight [**here**](https://www.reddit.com/r/rust/comments/rebk0r/comment/ho7osc8/?utm_source=share&utm_medium=web2x&context=3).

To summarize, they say that they are getting "not found" errors due to `assets/` directory being not present in the release archives:

```
thread 'main' panicked at 'File error: Couldn't load file ./assets/particle_effects.json: Error: IOError(Os { code: 2, kind: NotFound, message: "No such file or directory" })', src/resources.rs:465:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

What they propose is, either `assets/` directory should be bundled in binary and distributed that way or release archives should contain  `assets/`. Second option seems like something we should do anyways so I will work on that soon.

Another thing they mention is that binary cannot resolve the `assets/` path if not run in the same directory. For that, we can include a wrapper script that utilizes `FISHFIGHT_ASSETS`, very much like how the [AUR package](https://aur.archlinux.org/cgit/aur.git/tree/fishfight.sh?h=fishfight-git) is doing right now.

Tell me what you think 🐻

**Comment by olefasting** [comment]
Baking the assets directory into the binary is not really an option as it will require rebuilds if users modify some of the data, like when they add a map using the editor, for example. We will also have the same problem with the config file and a future mod directory...

The issue with running the executable from another directory is also a very common one that is somewhat hard to solve, unless you use the most common solutions; an env var, as you proposed, and/or OS-specific API's, like .desktop files on Linux or Documents and Settings and/or the registry, in the case of Windows.

If we want to use the env var, we will have to make some sort of install process, unless we want to instruct our users to set up their env manually, as it won't really solve the issue with running the binary with a non-standard working directory unless we use absolute paths. I think the best solution is to just instruct users to either run the game from the folder that holds the binary or to use `FISHFIGHT_ASSETS='/path' ./path/to/fishfight-bin` if they have to run it from somewhere else. Most of the end-users will probably be running the game on platforms like Steam, any ways, when release comes

We could maybe consider changing the `FISHFIGHT_ASSETS` var back to pointing to the root dir, instead of the assets directory, like it used to, so that we can use it to locate the config file and a future mod directory, as well.
If we do that, I think we should rename it to `FISHFIGHT_PATH` or something like that, for the sake of clarity, though :)

**Comment by woelfman** [comment]
Another option is to update `resources.rs` to use the executable path like so:
```rust
    let assets_dir_path = std::env::current_exe()?.parent().ok_or(ErrorKind::File)?.join(assets_dir);
```
A new `build.rs` would be appropriate for either linking or copying the `assets` directory to the target directory so that `cargo run` continues to work.

**Comment by olefasting** [comment]
That is also a possible solution. We should at least make sure to offer a way to circumvent it or, even better, not make it default behavior, as we never know if/when we might need correct/standard behavior and the possibility to use different working directories and have them as origin for relative paths...
After all, this problem really isn't with the game, as I see it, but the way most file systems and OS works, as the game is just acting as it is meant to, using the working directory as origin for relative paths, not the binary location (unless they are the same, of course)

**Comment by orhun** [comment]
To sum up our current status: we are now including the `assets` directory in release archives so users can simply run the fishfight binary in the same directory (after extracting the archive) and everything should be fine. But if the `assets` directory should be placed somewhere else, `FISHFIGHT_ASSETS` environment variable can be used to point to the correct location. This is not usually the case for regular users and it might be needed while packaging. So I'd like to create `PACKAGING.md` to mention this environment variable as well as the other details for distro packagers.

Other than that, I think our current behavior is pretty normal and I'm not sure what other actions we can take. (`build.rs` solution sounds fantastic though 😄 )

> We could maybe consider changing the `FISHFIGHT_ASSETS` var back to pointing to the root dir, instead of the assets directory, like it used to, so that we can use it to locate the config file and a future mod directory, as well. If we do that, I think we should rename it to `FISHFIGHT_PATH` or something like that, for the sake of clarity, though :)

Definitely!

**Comment by olefasting** [comment]
> Other than that, I think our current behavior is pretty normal and I'm not sure what other actions we can take. (`build.rs` solution sounds fantastic though smile )

I agree! Unsure how, for example, Steam handles installs but I am suspecting we will have to deal with working directories at some point..
I can do the stuff with the env variable in my current PR, which will be up within a weeks time, and I'll give you a heads up once it is up so that you can make the appropriate changes to the arch release
